src/normed.jl

```diff
@@ -25,77 +37,82 @@ for T in (UInt8, UInt16, UInt32, UInt64)
+zero(::Type{Normed{T,f}}) where {T <: Integer,f} = Normed{T,f}(zero(T),0)
+function oneunit(::Type{N}) where N <: Normed{T,f} where {T,f}
+    N(typemax(rawtype(N)) >> (8*sizeof(N)-nbitsfrac(N)-(T<:Signed)), 0)
 end
 one(::Type{T}) where {T <: Normed} = oneunit(T)
 zero(x::Normed) = zero(typeof(x))
 oneunit(x::Normed) =  one(typeof(x))
 one(x::Normed) = oneunit(x)
 rawone(v) = reinterpret(one(v))
 
-# Conversions
-function Normed{T,f}(x::Normed{T2}) where {T <: Unsigned,T2 <: Unsigned,f}
+# More construction and conversion
+function Normed{T,f}(x::Normed{T2}) where {T <: Integer,T2 <: Integer,f}
     U = Normed{T,f}
     y = round((rawone(U)/rawone(x))*reinterpret(x))
-    (0 <= y) & (y <= typemax(T)) || throw_converterror(U, x)
+    (typemin(T) <= y) & (y <= typemax(T)) || throw_converterror(U, x)
     reinterpret(U, _unsafe_trunc(T, y))
 end
 N0f16(x::N0f8) = reinterpret(N0f16, convert(UInt16, 0x0101*reinterpret(x)))
 
-(::Type{U})(x::Real) where {U <: Normed} = _convert(U, x)
+(::Type{N})(x::Real) where {N <: Normed} = _convert(N, x)
 
-function _convert(::Type{U}, x) where {T, f, U <: Normed{T,f}}
-    if T == UInt128 # for UInt128, we can't widen
+function _convert(::Type{N}, x) where {T <: Integer, f, N <: Normed{T,f}}
+    if T === UInt128 || T === Int128 # for [U]Int128, we can't widen
         # the upper limit is not exact
-        (0 <= x) & (x <= (typemax(T)/rawone(U))) || throw_converterror(U, x)
-        y = round(rawone(U)*x)
+        ((typemin(T)/rawone(N)) <= x) & (x <= (typemax(T)/rawone(N))) || throw_converterror(N, x)
+        y = round(rawone(N)*x)
     else
-        y = round(widen1(rawone(U))*x)
-        (0 <= y) & (y <= typemax(T)) || throw_converterror(U, x)
+        y = round(widen1(rawone(N))*x)
+        (typemin(T) <= y) & (y <= typemax(T)) || throw_converterror(N, x)
     end
-    reinterpret(U, _unsafe_trunc(T, y))
+    reinterpret(N, _unsafe_trunc(T, y))
 end
 # Prevent overflow (https://discourse.julialang.org/t/saving-greater-than-8-bit-images/6057)
-function _convert(::Type{U}, x::Float16) where {T, f, U <: Normed{T,f}}
-    if Float16(typemax(T)/rawone(U)) > Float32(typemax(T)/rawone(U))
-        x == Float16(typemax(T)/rawone(U)) && return typemax(U)
+function _convert(::Type{N}, x::Float16) where {T <: Integer, f, N <: Normed{T,f}}
+    if Float16(typemax(T)/rawone(N)) > Float32(typemax(T)/rawone(N))
+        x == Float16(typemax(T)/rawone(N)) && return typemax(N)
+        if T <: Signed
+            x == Float16(typemin(T)/rawone(N)) && return typemin(N)
+        end
     end
-    return _convert(U, Float32(x))
+    return _convert(N, Float32(x))
 end
-function _convert(::Type{U}, x::Tf) where {T, f, U <: Normed{T,f}, Tf <: Union{Float32, Float64}}
-    if T == UInt128 && f == 53
-        0 <= x <= Tf(3.777893186295717e22) || throw_converterror(U, x)
+function _convert(::Type{N}, x::Tf) where {T <: Integer, f, N <: Normed{T,f}, Tf <: Union{Float32, Float64}}
```

Comment:
I'm sorry but this method is currently optimized for non-negative numbers and cannot handle negative numbers. (pay attention to `k`.) Of course, this is a trivial issue, but I mean, "optimization is specialization".

Reply:
Yeah, I hadn't checked any of this, this is WIP.

Reply:
I don't care about that much *in `FixedPointNumbers`*. However, it is generally not good to ask other packages for such verification and modification. If this is *essential* to the design, we should change it aggressively, but I don't think so.

Reply:
Not sure I understand. All I meant was that this was a draft I decided to let people know I was working on. I will fix this before we merge this if we decide to do so.

Reply:
I think this is a simple problem. The *current* `Normed` is always unsigned, but your `Normed` is not always unsigned.